telegram/_bot.py

```diff
@@ -2037,13 +2046,51 @@ async def send_media_group(
                 :attr:`~telegram.request.BaseRequest.DEFAULT_NONE`.
             api_kwargs (:obj:`dict`, optional): Arbitrary keyword arguments to be passed to the
                 Telegram API.
+            caption (:obj:`str`, optional): Caption that will be added to the
+                first element of :paramref:`media`, so that it will be used as caption for the
+                whole media group.
+                Defaults to :obj:`None`.
+            parse_mode (:obj:`str` | :obj:`None`, optional):
+                Parse mode for :paramref:`caption`.
+                See the constants in :class:`telegram.constants.ParseMode` for the
+                available modes.
+            caption_entities (List[:class:`telegram.MessageEntity`], optional):
+                List of special entities for :paramref:`caption`,
+                which can be specified instead of :paramref:`parse_mode`.
+                Defaults to :obj:`None`.
 
         Returns:
             List[:class:`telegram.Message`]: An array of the sent Messages.
 
         Raises:
             :class:`telegram.error.TelegramError`
         """
+        if caption and any(
+            [
+                any(item.caption for item in media),
+                any(item.caption_entities for item in media),
+                # if parse_mode was set explicitly, even to None, error must be raised
+                any(item.parse_mode is not DEFAULT_NONE for item in media),
+            ]
+        ):
+            raise ValueError("You can only supply either group caption or media with captions.")
+
+        if caption:
+            # Create copies of objects in media and apply group caption to the first item.
+            # This will lead to the group being shown with this caption.
+
+            # Only copying of first item should be enough, but the behaviour of Defaults is such
+            # that a bot with a default parse mode will automatically set `parse_mode` for
+            # each item to default parse mode of the bot.
+            # This will throw ValueError at subsequent calls with same media items.
+            media = [copy.copy(item) for item in media]
```

Comment:
Uff, you're right. I think I stumbled about the "defaults-edits-things-in-place" (let's call it DETIP) issue somewhere before as well … Adding a workaround in `send_media_group` doesn't properly fix it though, instead one should edit `(Ext)Bot._insert_defaults` …

TBH, integrating `Defaults` into this PR has become more complicated than I had anticipated. IMO you've done a remarkable job wrapping your head around it and have prepared a very nice PR! Fixing DETIP would be somewhat out of scope for this PR IMO and also I'd feel bad for putting that on your plate as well 😬 

My suggestion would be: We put this PR on hold for a bit. In the meantime, I'll prepare a separate PR for fixing DETIP. Once that's done, we merge the changes into your branch & finish up this PR (which should hopefully be not too much work). Depending on when that happens, we can also see if you're still interested/available or if I finish up instead. Would that be okay with you?

In any case I want to thank you for the patience and work that you've put into the PR! I hope that you had some fun getting to know some of the internals of the library :)

Reply:
That is perfectly fine with me. Thanks for your support, and yes, it's been great fun, I've learned a lot already. Moving on to another issue I picked :)

Reply:
I just merged #3311. Once you merge the `master` branch into this branch, it should suffice to copy only the item that changes. Let me know if you're intrested in doing that yourself or if you'd rather have me finish up :)

Reply:
Thanks! I'm absolutely interested in trying to do this myself.